core/src/main/java/org/infinispan/util/concurrent/locks/impl/InfinispanLock.java

```diff
@@ -0,0 +1,537 @@
+    *
+    * @return {@code true} if the lock is acquired.
+    */
+   public boolean isLocked() {
+      return current != null;
+   }
+
+   /**
+    * It forces a deadlock checking.
+    */
+   public void deadlockCheck() {
+      LockPlaceHolder holder = current;
+      if (holder != null) {
+         for (LockPlaceHolder pending : pendingRequest) {
+            pending.checkDeadlock(holder.owner);
+         }
+      }
+   }
+
+   /**
+    * It tests if the lock has the lock owner.
+    * <p/>
+    * It return {@code true} if the lock owner is the current lock owner or it in the queue.
+    *
+    * @param lockOwner the lock owner to test.
+    * @return {@code true} if it contains the lock owner.
+    */
+   public boolean containsLockOwner(Object lockOwner) {
+      return lockOwners.containsKey(lockOwner);
+   }
+
+   private void onCanceled(LockPlaceHolder canceled) {
+      if (trace) {
+         log.tracef("Release lock for %s. It was canceled.", canceled.owner);
+      }
+      LockPlaceHolder wantToRelease = lockOwners.get(canceled.owner);
+      if (wantToRelease == null) {
+         if (trace) {
+            log.tracef("%s not found!", canceled.owner);
+         }
+         //nothing to release
+         return;
+      }
+      LockPlaceHolder currentLocked = current;
+      if (currentLocked == wantToRelease) {
+         if (!casRelease(currentLocked)) {
+            if (trace) {
+               log.tracef("Releasing lock for %s. It is the current lock owner but another thread changed it.", canceled.owner);
+            }
+            //another thread already released!
+         } else {
+            tryAcquire();
+         }
+      }
+   }
+
+   private boolean casAcquire(LockPlaceHolder lockPlaceHolder) {
+      return OWNER_UPDATER.compareAndSet(this, null, lockPlaceHolder);
+   }
+
+   private boolean casRelease(LockPlaceHolder lockPlaceHolder) {
+      return OWNER_UPDATER.compareAndSet(this, lockPlaceHolder, null);
+   }
+
+   private boolean remove(Object lockOwner) {
+      return lockOwners.remove(lockOwner) != null;
+   }
+
+   private void triggerReleased() {
+      if (releaseRunnable != null) {
+         releaseRunnable.run();
+      }
+   }
+
+   private void tryAcquire() {
+      do {
+         LockPlaceHolder nextPending = pendingRequest.peek();
+         if (trace) {
+            log.tracef("Try acquire. Next in queue=%s. Current=%s", nextPending, current);
+         }
+         if (nextPending == null) {
+            return;
+         }
+         if (casAcquire(nextPending)) {
+            //we set the current lock owner, so we must remove it from the queue
+            pendingRequest.remove(nextPending);
+            if (nextPending.setAvailable()) {
+               if (trace) {
+                  log.tracef("%s successfully acquired the lock.", nextPending);
+               }
+               //successfully acquired
+               deadlockCheck();
```

Comment:
Could we skip this? If we only do the deadlock checks from the timer task, then we can remove the `deadlockChecker` field and pass it as a parameter to `checkDeadlock`/`deadlockCheck` instead.